### PR TITLE
FEAT: Adding exclude modules param(#2044)

### PR DESCRIPTION
**Pull request by JINO-ROHIT**
This draft PR addresses the exclude modules parameter to keep specific layers out of target modules

Addresses #2044 

Changes made:

1. Changed LoraConfig class to take in exclude_modules param.
2. Added the exclusion logic within the check_target_module_exists method.

@BenjaminBossan Can you have a look if this is the right direction? This is my first time contributing at peft , would love to have your guidance here :)

### CODE REVIEW

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for taking on this feature. I haven't done an in-depth review yet, but from a quick look, I saw this:

1. Right now, you only added the option to the `LoraConfig`. However, since the logic is implemented on the `check_target_module_exists` function, actually all methods can benefit from this feature (not the prompt learning ones, as they don't have `target_modules`, but all the rest like LoKr, BOFT, IA³, etc.). Let's add the argument to all relevant configs, i.e. the config of each method that has `target_modules`.
2. We need to have tests for this new feature. Please check https://github.com/huggingface/peft/blob/main/tests/test_tuners_utils.py, which already contains a bunch of tests to check that the correct modules are targeted. Similar tests would need to be added for `exclude_modules`. Could you please take a look? Don't hesitate to ask if you have questions.

**Comment by JINO-ROHIT** [comment]
@BenjaminBossan thanks for the headsup, ill work on the points

**Comment by JINO-ROHIT** [comment]
@BenjaminBossan ive added the test cases too, lmk what you think

File: tests/test_tuners_utils.py [hunk 1]
```diff
@@ -415,6 +415,35 @@ def test_realistic_example(self):
         ]
         assert model.targeted_module_names == expected
 
+class TestExcludedModuleNames(unittest.TestCase):
```
Comment (BenjaminBossan) [inline tests/test_tuners_utils.py:418]: Can we get additional tests:

1. Let's add tests for `exclude_modules` being a list of str.
2. What happens if all targeted modules are excluded? I think a nice error message would be the right call in this situation.
3. What happens if we have excluded modules that don't exist? I think it's fine if some of the entries from `exclude_modules` don't match, as we also allow that for `target_modules`. But what if none of the entries in `exclude_modules` matches? I think this should also result in an error. WDYT?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for the latest changes. The tests are already shaping up quite nicely, but let's improve the test coverage a bit more (see my suggestion). 

Moreover, you missed a few configs still. These are all I could find, but please double-check:

- AdaLoraConfig
- BoftConfig
- VBLoRAConfig
- FourierFTConfig
- LNTuningConfig
- PolyConfig

File: tests/test_tuners_utils.py (hunk 1 above)
Comment (JINO-ROHIT) [reply tests/test_tuners_utils.py:418]: yeap makes sense, let me work through the tests and edge cases, also the missed out modules

Comment (JINO-ROHIT) [reply tests/test_tuners_utils.py:418]: Hey @BenjaminBossan i managed to add 1 and 2 cases. I wanted some help on how 3 should be done. As in, I see currently _check_target_module_exists is a private abstract method that every peft methods override and check for target modules existence. Now for exclude_modules, should we do the same? Or do i open a for loop within inject_adapter and check for keys there?

File: src/peft/tuners/tuners_utils.py [hunk 2]
```diff
@@ -914,6 +914,20 @@ def check_target_module_exists(config, key: str) -> bool | re.Match[str] | None:
         `bool` | `re.Match[str]` | `None`: True of match object if key matches any target modules from config, False or
         None if no match found
     """
+    if config.target_modules and config.exclude_modules:
+        if config.exclude_modules == config.target_modules:
```
Comment (BenjaminBossan) [inline src/peft/tuners/tuners_utils.py:918]: I don't think that this is a sufficient check. Just to give some examples:

- `target_modules` is a list of str and `exclude_modules` is a str, but both essentially target the same modules
- `target_modules` is `["lin0", "lin1", "lin2"]` and `exclude_modules` is `["lin0", "lin1"]`. They look different, but if `"lin2"` doesn't exist, they're actually the same

We thus need to check what modules were actually targeted, we can't rely on the passed arguments.

File: tests/test_tuners_utils.py (hunk 1 above)
Comment (BenjaminBossan) [reply tests/test_tuners_utils.py:418]: I don't think we need to go that far. If we check closer, we can see e.g. for LoRA:

https://github.com/huggingface/peft/blob/aa3bd8fbf6820bb000e148a289b6b0c48c339774/src/peft/tuners/lora/model.py#L158-L160

And if we check where that's used, it all comes back to this line:

https://github.com/huggingface/peft/blob/aa3bd8fbf6820bb000e148a289b6b0c48c339774/src/peft/tuners/tuners_utils.py#L486

So what would be an elegant way to implement some sanity checks here? I have one proposal, but maybe you have a better idea: Right now, `check_target_module_exists` simply returns a bool, i.e. `True` or `False` if the module matched. We could extend that and, say, return a special object if a module was excluded due to `exclude_modules`.

This special value could be a very simple object. Its `__bool__` should return `False` (so that users who rely on checking `if not self._check_target_module_exists(peft_config, key)` don't suddenly get different results).

When we call `_check_target_module_exists`, we can collect keys that did not match because of `exclude_modules` in one list and keys that did not match for other reasons in another list. When we exit the loop, if no key was matched because they were all matching `exclude_modules`, we can raise a nice error message to say so. Otherwise, we raise the existing error:

https://github.com/huggingface/peft/blob/aa3bd8fbf6820bb000e148a289b6b0c48c339774/src/peft/tuners/tuners_utils.py#L506-L510

On the other hand, if not a single key was excluded due to `exclude_modules`, even though `exclude_modules` was passed by the user, we can give a nice warning saying something like "You passed exclude_modules=[...] but no key was matching this".

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for your further work on this PR. You raised a good question about how to implement the checks I mentioned. I made a proposal, but please LMK if you have a better idea.

**Comment by JINO-ROHIT** [comment]
Hey, thanks for the suggestions. I think I fixed both the issues. Happy to work if more fixed needed, lmk

**Comment by JINO-ROHIT** [comment]
gotcha, ive made the changes and i think it looks okayish now

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Great refactor, thanks a lot, this looks very clean now. I only found a small issue, otherwise this looks good.

**Comment by JINO-ROHIT** [comment]
yeap, done now :)

**Comment by BenjaminBossan** [comment]
Please run `make style` :)

**Comment by JINO-ROHIT** [comment]
hmm weird, i actually ran make style already. is this because of windows? should i switch to a ubuntu and try again?

**Comment by BenjaminBossan** [comment]
I don't think it's a Windows issue. Do you perhaps have another ruff version installed? CI uses v0.6.8.

**Comment by JINO-ROHIT** [comment]
yeah i think its the version, hopefully it works now

**Comment by JINO-ROHIT** [comment]
my bad , i didnt notice the doc-builder wasnt installed, it should be alright now

**Comment by JINO-ROHIT** [comment]
i see the tests are failed due to a typeerror, this seems like an older python version run right?

**Comment by JINO-ROHIT** [comment]
sure lemme get on it

**Comment by JINO-ROHIT** [comment]
done!

**Comment by BenjaminBossan** [comment]
You really seem to be on the bad side of the CI, now there is an issue with ruff again. It should be easy enough to address, please check.

**Comment by JINO-ROHIT** [comment]
yeah seems like after adding the import, I had to lowercase the List Union[list[int], int]

**Comment by BenjaminBossan** [comment]
Of course that's not the end of it, now there is a merge conflict. It should be easy enough to fix but don't hesitate to ask if you have questions.

Thanks A TON for your patience.

**Comment by JINO-ROHIT** [comment]
hey, any clue why these tests are failing?

**Comment by BenjaminBossan** [comment]
Looks like it's unrelated, I'll investigate tomorrow.

**Comment by JINO-ROHIT** [comment]
alrighty

**Comment by JINO-ROHIT** [comment]
ahh okay, like this?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
The logic still doesn't work quite right but the good news is that we can simplify it.

Right now, we check `self.targeted_module_names` and also `is_target_modules_in_base_model`, which is a bit redundant. So let's get rid of the latter.

Additionally, to ensure that X-LoRA does not raise an error (since there, it's okay to not have target modules), we need to check that we're not using dummy target modules (a special value used by X-LoRA). I sketched this out and this diff makes the tests pass for me:

```diff
modified   src/peft/tuners/tuners_utils.py
@@ -437,13 +437,12 @@ class BaseTuner(nn.Module, ABC):
         peft_config = self._prepare_adapter_config(peft_config, model_config)
 
         self._prepare_model(peft_config, model)
-        is_target_modules_in_base_model = False
         key_list = [key for key, _ in model.named_modules()]
 
-        if getattr(peft_config, "target_modules", None) == DUMMY_TARGET_MODULES:
+        uses_dummy_target_modules = getattr(peft_config, "target_modules", None) == DUMMY_TARGET_MODULES
+        if uses_dummy_target_modules:
             # dummy adapter, we allow not matching any module
             key_list = []
-            is_target_modules_in_base_model = True
 
         # update peft_config.target_modules if required
         peft_config = _maybe_include_all_linear_layers(peft_config, model)
@@ -494,19 +493,12 @@ class BaseTuner(nn.Module, ABC):
                 unmatched_modules.append(key)
             else:
                 self.targeted_module_names.append(key)
-                is_target_modules_in_base_model = True
                 parent, target, target_name = _get_submodules(model, key)
                 ctx = init_empty_weights if low_cpu_mem_usage else nullcontext
                 with ctx():
                     self._create_and_replace(peft_config, adapter_name, target, target_name, parent, current_key=key)
 
-        # Handle X-LoRA case.
-        if not is_target_modules_in_base_model and hasattr(peft_config, "target_modules"):
-            raise ValueError(
-                f"Target modules {peft_config.target_modules} not found in the base model. "
-                f"Please check the target modules and try again."
-            )
-        elif not self.targeted_module_names:
+        if not self.targeted_module_names and not uses_dummy_target_modules:
             if excluded_modules and not unmatched_modules:
                 # All targeted modules were excluded
                 raise ValueError(
```

**Comment by JINO-ROHIT** [comment]
thanks so much for the detailed suggestion!

**Comment by JINO-ROHIT** [comment]
some unrelated tests failing right? i think with windows some tensor equality is not fulfilled?

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Yes, CI is failing for unrelated reasons. I'll just re-run until the flaky test passes :)

So this PR is basically finished, but when I did a final review, I saw a minor oversight, please check my comment. But then we should really be good to go!

**Comment by JINO-ROHIT** [comment]
yeap done! thanks for guiding all the way through!

**Review by BenjaminBossan — APPROVED** [review]
Thanks for this nice PR, good work.
